examples/gpt-2/main.cpp

```diff
@@ -703,11 +796,16 @@ bool gpt2_eval(
     ggml_allocr_alloc_graph(allocr, gf);
 
     // run the computation
-    struct ggml_cplan plan = ggml_graph_plan(gf, n_threads);
-    static std::vector<uint8_t> work_buffer;
-    work_buffer.resize(plan.work_size);
-    plan.work_data = work_buffer.data();
-    ggml_graph_compute(gf, &plan);
+    if (ggml_backend_is_cpu(model.backend)) {
+        ggml_backend_cpu_set_n_threads(model.backend, n_threads);
+    }
+#ifdef GGML_USE_METAL
+    // TODO: not great - what should we do?
+    if (ggml_backend_is_metal(model.backend)) {
+        ggml_backend_metal_set_n_threads(model.backend, n_threads);
+    }
+#endif
```

Comment:
@slaren Any ideas how to avoid this `ifdef`.  It's not terrible, but maybe try to minimize such branching from the start

Reply:
In the long term, I would like to remove all `#ifdefs` entirely and just make the backend init functions return `NULL` if the backend is not available. That should allow using a single build for all backends. But the checks before calling a backend-specific function will always be required.

Reply:
We could extend the backend API with `set_n_threads()` and backends like CUDA would be just noops.
This specific code would become a one line without `#ifdef`

```c++
ggml_backend_set_n_threads(model.backend, n_threads);
```

Just an idea - we can just leave it as it is

Reply:
We can't do this for every backend-specific function. I am not sure that calling this function `ggml_backend_metal_set_n_threads` is very honest here. If it was renamed to `ggml_backend_metal_set_n_cb`, the problem would become more obvious.

Reply:
Yes, it's better to have the correct name